File: backend/app/api/user_settings.py
"""
用户配置管理API
基于实际功能需求设计的配置系统
"""
from fastapi import APIRouter, HTTPException
from sqlalchemy.orm import Session
from pydantic import BaseModel
from typing import Optional, List

from app.database import get_db
from app.services.config_service import ConfigService

router = APIRouter(prefix="/api/v1/settings", tags=["用户配置"])


# ===================== 配置数据模型 =====================

class AnalysisConfig(BaseModel):
    """分析配置"""
    default_timeframe: str = "daily"  # 默认分析周期
    multi_timeframes: List[str] = ["daily", "weekly", "30", "60"]  # 多周期默认启用
    signal_threshold: int = 3  # 信号触发阈值（分数）
    enable_cache: bool = True  # 是否启用缓存


class DataConfig(BaseModel):
    """数据配置"""
    kline_count: int = 500  # K线数量
    cache_ttl_hours: int = 1  # 缓存时长（小时）
    enable_redis: bool = True  # 是否启用Redis


class DisplayConfig(BaseModel):
    """显示配置"""
    watchlist_columns: int = 5  # 关注列表每行显示数量
    default_sort: str = "score_desc"  # 默认排序
    show_investment_advice: bool = True  # 显示投资建议


class NotificationConfig(BaseModel):
    """通知配置"""
    feishu_webhook: Optional[str] = None  # 飞书Webhook URL
    enable_notification: bool = False  # 是否启用通知
    min_notify_score: int = 4  # 最小通知评分
    rate_limit_minutes: int = 30  # 限流时间（分钟）


class TradingConfig(BaseModel):
    """交易建议配置"""
    stop_loss_percent: float = 5.0  # 止损建议（%）
    take_profit_percent: float = 8.0  # 止盈建议（%）
    position_percent: float = 20.0  # 建议仓位（%）


class AllSettings(BaseModel):
    """所有配置"""
    analysis: AnalysisConfig
    data: DataConfig
    display: DisplayConfig
    notification: NotificationConfig
    trading: TradingConfig


# ===================== 配置存储 =====================

# 简单内存存储（生产环境应使用数据库）
_settings_storage = {
    "analysis": {
        "default_timeframe": "daily",
        "multi_timeframes": ["daily", "weekly", "30", "60"],
        "signal_threshold": 3,
        "enable_cache": True
    },
    "data": {
        "kline_count": 500,
        "cache_ttl_hours": 1,
        "enable_redis": True
    },
    "display": {
        "watchlist_columns": 5,
        "default_sort": "score_desc",
        "show_investment_advice": True
    },
    "notification": {
        "feishu_webhook": None,
        "enable_notification": False,
        "min_notify_score": 4,
        "rate_limit_minutes": 30
    },
    "trading": {
        "stop_loss_percent": 5.0,
        "take_profit_percent": 8.0,
        "position_percent": 20.0
    }
}


# ===================== API接口 =====================

@router.get("/")
async def get_all_settings():
    """获取所有配置"""
    return _settings_storage


@router.get("/{category}")
async def get_settings_category(category: str):
    """获取指定类别的配置"""
    if category not in _settings_storage:
        raise HTTPException(status_code=404, detail=f"配置类别不存在: {category}")
    return _settings_storage[category]


@router.put("/{category}")
async def update_settings_category(category: str, settings: dict):
    """更新指定类别的配置"""
    if category not in _settings_storage:
        raise HTTPException(status_code=404, detail=f"配置类别不存在: {category}")

    # 更新配置
    _settings_storage[category].update(settings)

    return {
        "message": f"{category} 配置已更新",
        "settings": _settings_storage[category]
    }


@router.post("/")
async def save_all_settings(settings: AllSettings):
    """保存所有配置"""
    _settings_storage["analysis"] = settings.analysis.dict()
    _settings_storage["data"] = settings.data.dict()
    _settings_storage["display"] = settings.display.dict()
    _settings_storage["notification"] = settings.notification.dict()
    _settings_storage["trading"] = settings.trading.dict()

    return {
        "message": "所有配置已保存",
        "settings": _settings_storage
    }


@router.post("/reset")
async def reset_settings():
    """重置所有配置为默认值"""
    global _settings_storage
    _settings_storage = {
        "analysis": {
            "default_timeframe": "daily",
            "multi_timeframes": ["daily", "weekly", "30", "60"],
            "signal_threshold": 3,
            "enable_cache": True
        },
        "data": {
            "kline_count": 500,
            "cache_ttl_hours": 1,
            "enable_redis": True
        },
        "display": {
            "watchlist_columns": 5,
            "default_sort": "score_desc",
            "show_investment_advice": True
        },
        "notification": {
            "feishu_webhook": None,
            "enable_notification": False,
            "min_notify_score": 4,
            "rate_limit_minutes": 30
        },
        "trading": {
            "stop_loss_percent": 5.0,
            "take_profit_percent": 8.0,
            "position_percent": 20.0
        }
    }

    return {
        "message": "配置已重置为默认值",
        "settings": _settings_storage
    }

File: frontend/user_settings.js
// ===================== 用户配置管理函数 =====================

// 加载所有配置
async function loadAllSettings() {
    try {
        const response = await fetch(`${API_BASE}/api/v1/settings/`);
        if (!response.ok) throw new Error('加载配置失败');

        const settings = await response.json();

        // 分析配置
        if (settings.analysis) {
            document.getElementById('setting_default_timeframe').value = settings.analysis.default_timeframe || 'daily';
            document.getElementById('setting_signal_threshold').value = settings.analysis.signal_threshold || 3;
            document.getElementById('setting_enable_cache').checked = settings.analysis.enable_cache !== false;

            // 多周期复选框
            const timeframes = settings.analysis.multi_timeframes || ['daily', 'weekly', '30', '60'];
            document.getElementById('setting_tf_30').checked = timeframes.includes('30');
            document.getElementById('setting_tf_60').checked = timeframes.includes('60');
            document.getElementById('setting_tf_daily').checked = timeframes.includes('daily');
            document.getElementById('setting_tf_weekly').checked = timeframes.includes('weekly');
            document.getElementById('setting_tf_monthly').checked = timeframes.includes('monthly');
        }

        // 数据配置
        if (settings.data) {
            document.getElementById('setting_kline_count').value = settings.data.kline_count || 500;
            document.getElementById('setting_cache_ttl').value = settings.data.cache_ttl_hours || 1;
            document.getElementById('setting_enable_redis').checked = settings.data.enable_redis !== false;
        }

        // 显示配置
        if (settings.display) {
            document.getElementById('setting_watchlist_columns').value = settings.display.watchlist_columns || 5;
            document.getElementById('setting_default_sort').value = settings.display.default_sort || 'score_desc';
            document.getElementById('setting_show_advice').checked = settings.display.show_investment_advice !== false;
        }

        // 通知配置
        if (settings.notification) {
            document.getElementById('setting_feishu_webhook').value = settings.notification.feishu_webhook || '';
            document.getElementById('setting_min_notify_score').value = settings.notification.min_notify_score || 4;
            document.getElementById('setting_rate_limit').value = settings.notification.rate_limit_minutes || 30;
            document.getElementById('setting_enable_notification').checked = settings.notification.enable_notification || false;
        }

        // 交易建议配置
        if (settings.trading) {
            document.getElementById('setting_stop_loss').value = settings.trading.stop_loss_percent || 5;
            document.getElementById('setting_take_profit').value = settings.trading.take_profit_percent || 8;
            document.getElementById('setting_position').value = settings.trading.position_percent || 20;
        }

        addLog('配置加载', '✅ 配置已加载');

    } catch (error) {
        console.error('加载配置失败:', error);
        addLog('配置加载失败', error.message);
        alert(`❌ 加载配置失败: ${error.message}`);
    }
}

// 保存所有配置
async function saveAllSettings() {
    try {
        // 收集表单数据
        const settings = {
            analysis: {
                default_timeframe: document.getElementById('setting_default_timeframe').value,
                signal_threshold: parseInt(document.getElementById('setting_signal_threshold').value),
                enable_cache: document.getElementById('setting_enable_cache').checked,
                multi_timeframes: []
            },
            data: {
                kline_count: parseInt(document.getElementById('setting_kline_count').value),
                cache_ttl_hours: parseFloat(document.getElementById('setting_cache_ttl').value),
                enable_redis: document.getElementById('setting_enable_redis').checked
            },
            display: {
                watchlist_columns: parseInt(document.getElementById('setting_watchlist_columns').value),
                default_sort: document.getElementById('setting_default_sort').value,
                show_investment_advice: document.getElementById('setting_show_advice').checked
            },
            notification: {
                feishu_webhook: document.getElementById('setting_feishu_webhook').value || null,
                min_notify_score: parseInt(document.getElementById('setting_min_notify_score').value),
                rate_limit_minutes: parseInt(document.getElementById('setting_rate_limit').value),
                enable_notification: document.getElementById('setting_enable_notification').checked
            },
            trading: {
                stop_loss_percent: parseFloat(document.getElementById('setting_stop_loss').value),
                take_profit_percent: parseFloat(document.getElementById('setting_take_profit').value),
                position_percent: parseFloat(document.getElementById('setting_position').value)
            }
        };

        // 收集多周期复选框
        if (document.getElementById('setting_tf_30').checked) settings.analysis.multi_timeframes.push('30');
        if (document.getElementById('setting_tf_60').checked) settings.analysis.multi_timeframes.push('60');
        if (document.getElementById('setting_tf_daily').checked) settings.analysis.multi_timeframes.push('daily');
        if (document.getElementById('setting_tf_weekly').checked) settings.analysis.multi_timeframes.push('weekly');
        if (document.getElementById('setting_tf_monthly').checked) settings.analysis.multi_timeframes.push('monthly');

        const response = await fetch(`${API_BASE}/api/v1/settings/`, {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify(settings)
        });

        if (!response.ok) throw new Error('保存配置失败');

        addLog('配置保存', '✅ 配置已保存');
        alert('✅ 配置保存成功！部分配置需要刷新页面后生效。');

    } catch (error) {
        console.error('保存配置失败:', error);
        addLog('配置保存失败', error.message);
        alert(`❌ 保存配置失败: ${error.message}`);
    }
}

// 重置所有配置
async function resetAllSettings() {
    if (!confirm('确定要重置所有配置为默认值吗？')) return;

    try {
        const response = await fetch(`${API_BASE}/api/v1/settings/reset`, {
            method: 'POST'
        });

        if (!response.ok) throw new Error('重置配置失败');

        await loadAllSettings();
        addLog('配置重置', '✅ 配置已重置为默认值');

    } catch (error) {
        console.error('重置配置失败:', error);
        addLog('配置重置失败', error.message);
        alert(`❌ 重置配置失败: ${error.message}`);
    }
}

// 应用配置到前端功能
function applySettings() {
    // 从localStorage或其他存储获取配置并应用
    const watchlistColumns = localStorage.getItem('setting_watchlist_columns');
    if (watchlistColumns) {
        // 可以动态调整关注列表布局
        console.log('应用显示配置:', watchlistColumns);
    }
}
